### PR TITLE
Refresh primary clone concurrently

### PR DESCRIPTION
**Pull request by listx**
That is, only lock the primary clone when we first create it (and when we check for its existence as part of the "does the primary exist" check).

This is blocked by #30294 

/cc @cjwagner @airbornepony @timwangmusic 
/hold

### CODE REVIEW

File: prow/git/v2/client_factory.go
Comment (listx) [reply prow/git/v2/client_factory.go:396]: The reason I didn't do this is because we still hold the lock for calling RemoteUpdate().

But, I agree that having a single defer is easier to read. I could create the maybeCreatePrimary function and put the call to RemoteUpdate() in there. But then a better name would probably be maybeCloneAndUpdatePrimary. WDYT?

Comment (airbornepony) [reply prow/git/v2/client_factory.go:396]: Ah, I see that FetchCommits is the alternative to RemoteUpdate(). So you could do like you said and gate the FetchCommits outside the call. I think this would be easier to read and maintain still.

File: prow/git/v2/client_factory.go
Comment (airbornepony) [inline prow/git/v2/client_factory.go:381]: Since RemoteUpdate() uses "--prune", probably should document on RepoOpts.ShareObjectsWithSourceRepo that it must be used only in conjunction with FetchCommits. (This is preexisting behavior, but important to understand for future use/review of these features.)

**Comment by listx** [comment]
This PR is blocked by https://github.com/kubernetes/test-infra/pull/30294. Once that merges (with its integration test), we should be confident about how the changes in this PR behave under heavy traffic.

File: prow/git/v2/client_factory.go
Comment (listx) [reply prow/git/v2/client_factory.go]: Done.

File: prow/git/v2/client_factory.go
Comment (listx) [reply prow/git/v2/client_factory.go]: Done (also added a check to enforce this rule before we create the primary/secondary clones).

**Review by airbornepony — APPROVED** [review]

**Comment by listx** [comment]
This is ready for review. The integration test has been updated to be a little more strict as well (we now check that each of the 100 PRs have the expected (unique) job names).

File: prow/git/v2/client_factory.go
Comment (cjwagner) [inline prow/git/v2/client_factory.go:425]: For this functionality to work as expected, I think this assumes that `repoOpts.ShareObjectsWithSourceRepo == true` which is not necessarily true at this point. If we're not sharing objects with the source repo then this needs to somehow update the secondary repo, not just the primary, right?

Comment (listx) [reply prow/git/v2/client_factory.go:425]: No, because we are fetching new objects into the primary. It doesn't matter if we're sharing or not sharing across the primary or secondary because:

- if we're sharing, the secondary doesn't need to fetch the same objects (redundantly) again, because the primary has it already
- if we're not sharing, the secondary will get a full copy of everything in the (updated) primary that already has the objects

IOW the primary is *always created/refreshed first* before the secondary is created from it (either with a partial or non-partial (full) checkout).

**Review by cjwagner — APPROVED** [review]

**Pull request merged**
